src/AutoRest.CSharp.V3/Output/Builders/SerializationBuilder.cs

```diff
@@ -150,26 +153,82 @@ private static XmlObjectSerialization BuildXmlObjectSerialization(ObjectSchema o
                 schemaTypeReference, elements.ToArray(), attributes.ToArray(), embeddedArrays.ToArray());
         }
 
+        private static JsonPropertySerialization CreateJsonPropertySerialization(Property property) => new JsonPropertySerialization(
+            property.SerializedName,
+            property.CSharpName(),
+            BuildSerialization(property.Schema, property.IsNullable(), property.IsBag),
+            property.IsBag
+        );
+
         private static JsonObjectSerialization BuildJsonObjectSerialization(ObjectSchema objectSchema, TypeReference schemaTypeReference, bool isNullable)
         {
             List<JsonPropertySerialization> serializationProperties = new List<JsonPropertySerialization>();
             foreach (var schema in EnumerateHierarchy(objectSchema))
             {
-                foreach (Property property in schema.Properties!)
+                foreach (Property property in AggregateProperties(schema))
                 {
-                    JsonPropertySerialization propertySerialization = new JsonPropertySerialization(
-                        property.SerializedName,
-                        property.CSharpName(),
-                        BuildSerialization(property.Schema, property.IsNullable())
-                    );
-
-                    serializationProperties.Add(propertySerialization);
+                    serializationProperties.Add(CreateJsonPropertySerialization(property));
                 }
             }
 
             return new JsonObjectSerialization(schemaTypeReference, serializationProperties.ToArray(), CreateAdditionalProperties(objectSchema));
         }
 
+        private class PropertyBag
+        {
+            public Dictionary<string, PropertyBag> Bag { get; } = new Dictionary<string, PropertyBag>();
+            public List<Property> Properties { get; } = new List<Property>();
+        }
+
+        private static IEnumerable<Property> AggregateProperties(ObjectSchema schema)
+        {
+            PropertyBag propertyBag = new PropertyBag();
+            propertyBag.Properties.AddRange(schema.Properties ?? Array.Empty<Property>());
+            PopulatePropertyBag(propertyBag, 0);
+            return propertyBag.Properties;
+        }
+
+        private static Property CreatePropertyFromPropertyBag(string name, PropertyBag bag) => new Property
```

Comment:
Let's try to implement this without synthesizing new code model types.

Reply:
Done.